rust/src/datatypes.rs

```diff
@@ -68,44 +69,80 @@ pub struct Field {
     nullable: bool,
 }
 
+pub trait ArrowNativeType: Send + Sync + Copy + PartialOrd + FromStr + 'static {}
+
 /// Trait indicating a primitive fixed-width type (bool, ints and floats).
-///
-/// This trait is a marker trait to indicate a primitive type, i.e. a type that occupies a fixed
-/// size in memory as indicated in bit or byte width.
-pub trait ArrowPrimitiveType: Send + Sync + Copy + PartialOrd + 'static {}
-
-impl ArrowPrimitiveType for bool {}
-impl ArrowPrimitiveType for u8 {}
-impl ArrowPrimitiveType for u16 {}
-impl ArrowPrimitiveType for u32 {}
-impl ArrowPrimitiveType for u64 {}
-impl ArrowPrimitiveType for i8 {}
-impl ArrowPrimitiveType for i16 {}
-impl ArrowPrimitiveType for i32 {}
-impl ArrowPrimitiveType for i64 {}
-impl ArrowPrimitiveType for f32 {}
-impl ArrowPrimitiveType for f64 {}
+pub trait ArrowPrimitiveType: 'static {
+    /// Corresponding native type for the primitive type.
+    type T: ArrowNativeType;
```

Comment:
Could You please use a more meaningful associated type name e.g. Native? For example `Option<T::T>` looks pretty weird.